parsers/typescript.c

```diff
@@ -0,0 +1,2001 @@
+		result->status = PARSER_FINISHED;
+		result->unusedChars = 1;
+		return;
+	}
+
+	result->status = PARSER_FAILED;
+}
+
+PARSER_DEF (AsyncKeyword, parseWord, "async", (int *))
+PARSER_DEF (ClassKeyword, parseWord, "class", (int *))
+PARSER_DEF (ConstKeyword, parseWord, "const", (int *))
+PARSER_DEF (ConstructorKeyword, parseWord, "constructor", (int *))
+PARSER_DEF (EnumKeyword, parseWord, "enum", (int *))
+PARSER_DEF (ForKeyword, parseWord, "for", (int *))
+PARSER_DEF (FunctionKeyword, parseWord, "function", (int *))
+PARSER_DEF (InterfaceKeyword, parseWord, "interface", (int *))
+PARSER_DEF (LetKeyword, parseWord, "let", (int *))
+PARSER_DEF (NamespaceKeyword, parseWord, "namespace", (int *))
+PARSER_DEF (OfKeyword, parseWord, "of", (int *))
+PARSER_DEF (PrivateKeyword, parseWord, "private", (int *))
+PARSER_DEF (ProtectedKeyword, parseWord, "protected", (int *))
+PARSER_DEF (PublicKeyword, parseWord, "public", (int *))
+PARSER_DEF (ReadonlyKeyword, parseWord, "readonly", (int *))
+PARSER_DEF (StaticKeyword, parseWord, "static", (int *))
+PARSER_DEF (ThisKeyword, parseWord, "this", (int *))
+PARSER_DEF (TypeKeyword, parseWord, "type", (int *))
+PARSER_DEF (VarKeyword, parseWord, "var", (int *))
+PARSER_DEF (WhileKeyword, parseWord, "while", (int *))
+
+SINGLE_CHAR_PARSER_DEF (Semicolon, ';', TOKEN_SEMICOLON)
+SINGLE_CHAR_PARSER_DEF (Comma, ',', TOKEN_COMMA)
+SINGLE_CHAR_PARSER_DEF (Colon, ':', TOKEN_COLON)
+SINGLE_CHAR_PARSER_DEF (Period, '.', TOKEN_PERIOD)
+SINGLE_CHAR_PARSER_DEF (OpenCurly, '{', TOKEN_OPEN_CURLY)
+SINGLE_CHAR_PARSER_DEF (CloseCurly, '}', TOKEN_CLOSE_CURLY)
+SINGLE_CHAR_PARSER_DEF (OpenParen, '(', TOKEN_OPEN_PAREN)
+SINGLE_CHAR_PARSER_DEF (CloseParen, ')', TOKEN_CLOSE_PAREN)
+SINGLE_CHAR_PARSER_DEF (OpenSquare, '[', TOKEN_OPEN_SQUARE)
+SINGLE_CHAR_PARSER_DEF (CloseSquare, ']', TOKEN_CLOSE_SQUARE)
+SINGLE_CHAR_PARSER_DEF (EqualSign, '=', TOKEN_EQUAL_SIGN)
+SINGLE_CHAR_PARSER_DEF (Star, '*', TOKEN_STAR)
+SINGLE_CHAR_PARSER_DEF (At, '@', TOKEN_AT)
+SINGLE_CHAR_PARSER_DEF (NewLine, '\n', TOKEN_NL)
+SINGLE_CHAR_PARSER_DEF (QuestionMark, '?', TOKEN_QUESTION_MARK)
+SINGLE_CHAR_PARSER_DEF (EOF, EOF, TOKEN_EOF)
+SINGLE_CHAR_PARSER_DEF (Pipe, '|', TOKEN_PIPE)
+SINGLE_CHAR_PARSER_DEF (Ampersand, '&', TOKEN_AMPERSAND)
+
+WORD_TOKEN_PARSER_DEF (Arrow, "=>", TOKEN_ARROW)
+
+PARSER_DEF (StringSQuote, parseString, '\'', (char *))
+PARSER_DEF (StringDQuote, parseString, '"', (char *))
+PARSER_DEF (StringTemplate, parseString, '`', (char *))
+
+BLOCK_PARSER_DEF (Parens, '(', ')', TOKEN_PARENS)
+BLOCK_PARSER_DEF (Squares, '[', ']', TOKEN_SQUARES)
+BLOCK_PARSER_DEF (Template, '<', '>', TOKEN_TEMPLATE)
+BLOCK_PARSER_DEF (Curlies, '{', '}', TOKEN_CURLIES)
+
+static bool tryParser(Parser parser, ParserStateInit stInit, ParserStateFree stFree, tokenInfo *const token)
+{
+	void *currentState = NULL;
+	parserResult result;
+	result.status = PARSER_NEEDS_MORE_INPUT;
+	result.unusedChars = 0;
+	charArray *usedC = charArrayNew ();
+	char c;
```

Comment:
int is needed for c.

Functions for reading input returns int, not char. There is a reason.
These function may return 0  ~ 255 or EOF.
If the platform that we use uses singned char as "char", we cannot receive value 255. 

```
[yamato@slave]/tmp% cat foo.c
cat foo.c
#include <stdio.h>

int f()
{
  return 255;
}

int e()
{
  return EOF;
}

int main(void)
{
  signed char cs0   = f();
  signed char cs1   = e();  
  printf("255: %d, EOF: %d\n", (int) cs0, (int)cs1);
  return 0;
}
[yamato@slave]/tmp% gcc foo.c
gcc foo.c
[yamato@slave]/tmp% ./a.out 
./a.out 
255: -1, EOF: -1
```
The parser cannot distinguish 255 and EOF.

If the platform that we use uses unsigned char as "char", we cannot receive value 255. 
```
[yamato@slave]/tmp% cat bar.c
cat bar.c
#include <stdio.h>

int f()
{
  return 255;
}

int e()
{
  return EOF;
}

int main(void)
{
  unsigned char cs0   = f();
  unsigned char cs1   = e();  
  printf("255: %d, EOF: %d\n", (int)cs0, (int)cs1);
  return 0;
}
[yamato@slave]/tmp% gcc bar.c
gcc bar.c
[yamato@slave]/tmp% ./a.out 
./a.out 
255: 255, EOF: 255
```

Either case, int is needed to receive the value returned from the input functions.
See https://github.com/universal-ctags/ctags/pull/1697